### PR TITLE
Fix #7: Add fuzzer, fix utf-8 validation, disallow `frame_size` of 0.

### LINKED ISSUE #7: Fuzzing
I think it would be good to create a fuzzer for the decoder to verify it doesn't panic or UB when sent random gibberish.

### PR DESCRIPTION
**Pull request by finnbear**
## Changes
- Added a fuzzer for receiving arbitrary data (`cargo fuzz run decode`)
- Fixed an instance of UB in utf-8 validation, found by fuzzer (I didn't see a security policy)
- Panic if `Config.frame_size` is `0` to avoid OOM, found by fuzzer
- Add 15m of fuzzing to CI

## Related

Fixes #7

### CODE REVIEW

File: fuzz/fuzz_targets/decode.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
+#![no_main]
+use std::task::{Context, Poll};
+
+use libfuzzer_sys::fuzz_target;
+use std::io;
+use futures::stream::StreamExt;
+use std::pin::Pin;
+extern crate tokio_websockets;
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
```
Comment (Gelbpunkt) [inline fuzz/fuzz_targets/decode.rs:9]: Could you please format these using the main rustfmt.toml? I assume symlinking it in the fuzz subdirectory should be sufficient.

File: fuzz/fuzz_targets/decode.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    let mut ws = tokio_websockets::ServerBuilder::new().serve(DeterministicStream{
+        reads
+    });
+
+    futures::executor::block_on(async move {
+        for _ in 0..10 {
+            let _ = ws.next().await;
+        }
```
Comment (Gelbpunkt) [inline fuzz/fuzz_targets/decode.rs:64]: I'm not too familiar with fuzzing, so this seems a bit arbitrary to me. Why would we try to read exactly 10 messages?

**Review by Gelbpunkt — CHANGES_REQUESTED** [review]
Thanks for tackling this! I'm super happy that the value of fuzzing is already showing with the UTF-8 validation catch.

File: fuzz/fuzz_targets/decode.rs (hunk 1 above)
Comment (finnbear) [reply fuzz/fuzz_targets/decode.rs]: Done; no symlink required as `rustfmt.toml` in any parent directory works (https://github.com/rust-lang/rustfmt/blob/182a20378b3a844e3677ff93d96f8ac70b9f009a/Configurations.md?plain=1#L3)

File: fuzz/fuzz_targets/decode.rs (hunk 2 above)
Comment (finnbear) [reply fuzz/fuzz_targets/decode.rs]: 10 should have been enough to catch all errors, while ensuring a finite duration. However, I've replaced this with an `Arbitrary` amount of operations.

**Comment by finnbear** [comment]
I tried using `act` to debug the workflow but it didn't work. Consider disabling the approval requirement until I finish debugging via GitHub ;)

**Comment by finnbear** [comment]
Ready for re-review :)

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -39,6 +39,34 @@ jobs:
...
+
+      - name: Install cargo-fuzz
+        uses: baptiste0928/cargo-install@v3
+        with:
+          crate: cargo-fuzz
+          locked: false
+
+      - name: Run stream fuzzer for a limited time
+        run: |
+          RUST_BACKTRACE=1 cargo fuzz run stream -- -max_total_time=60
```
Comment (Gelbpunkt) [inline .github/workflows/ci.yml:69]: I think it should be fine to run for longer than a minute, the powerset CI already runs for a long time anyways

File: fuzz/fuzz_targets/stream.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+};
+
+use futures::{stream::StreamExt, SinkExt};
+use libfuzzer_sys::fuzz_target;
+use tokio_websockets::{CloseCode, Config, Limits, Message};
+extern crate tokio_websockets;
+use std::convert::TryFrom;
+
+use arbitrary::Arbitrary;
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
```
Comment (Gelbpunkt) [inline fuzz/fuzz_targets/stream.rs:16]: The import ordering is still completely different than in the rustfmt.toml. It also would be good if the CI step for rustfmt would catch this :wink:

**Review by Gelbpunkt — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 3 above)
Comment (finnbear) [reply .github/workflows/ci.yml]: Since powerset takes ~15m, I raised fuzzing to 15m

File: fuzz/fuzz_targets/stream.rs (hunk 4 above)
Comment (finnbear) [reply fuzz/fuzz_targets/stream.rs]: This was the output of `cargo fmt`. I think `extern crate` in the middle was resetting the ordering. Moving `extern crate` above the imports seems to have resolved it.

Comment (Gelbpunkt) [reply fuzz/fuzz_targets/stream.rs]: Are you using nightly rustfmt? The rustfmt.toml uses some nightly-exclusive options to make the formatting more uniform

Comment (finnbear) [reply fuzz/fuzz_targets/stream.rs]: I use `rustc 1.79.0-nightly (f9b161492 2024-04-19)`. Is there still a formatting problem?

Comment (Gelbpunkt) [reply fuzz/fuzz_targets/stream.rs]: No, the formatting is fine now :)

**Review by Gelbpunkt — APPROVED** [review]

**Review by vilgotf — APPROVED** [review]
Noice

**Pull request merged**